common/scala/src/main/scala/whisk/core/database/DocumentFactory.scala

```diff
@@ -166,24 +166,29 @@ trait DocumentFactory[W] extends MultipleReadersSingleWriterCache[W, DocInfo] {
 
   def attach[Wsuper >: W](
     db: ArtifactStore[Wsuper],
-    doc: DocInfo,
+    doc: W,
+    docInfo: DocInfo,
     attachmentName: String,
     contentType: ContentType,
     bytes: InputStream)(implicit transid: TransactionId, notifier: Option[CacheChangeNotification]): Future[DocInfo] = {
 
     Try {
       require(db != null, "db undefined")
       require(doc != null, "doc undefined")
+      require(docInfo != null, "docInfo undefined")
     } map { _ =>
       implicit val logger = db.logging
       implicit val ec = db.executionContext
 
-      val key = CacheKey(doc.id.asDocInfo)
-      // invalidate the key because attachments update the revision;
-      // do not cache the new attachment (controller does not need it)
-      cacheInvalidate(key, {
-        val src = StreamConverters.fromInputStream(() => bytes)
-        db.attach(doc, attachmentName, contentType, src)
+      val key = CacheKey(doc)
```

Comment:
the `key` for `doc: WhiskEntity` is `docInfo`, so you could create a `DocInfo` instance out of the key.
alternatively, if `W` is bounded to `WhiskDocument` then you can get its `id.asDocInfo`.

Reply:
The bound is now added to the trait.